analysis toolbox: improve latest_data() method (e.g., allow retrieving latest n measurement timestamps and folders)

chellings opened the pull request:
UPDATE: Instead of adding Nathan's function, the pull request now modifies latest_data() in the analysis toolbox to implement the same functionality. See discussion below.

A nice little helper function by @nathlacroix 

This new static function in base_analysis should not interfere with anything else.

Inviting @stephlazar to review
FYI @antsr

nathlacroix committed:
Add functionality to base analysis

Add function to retrieve latest n measurement timestamps and folders

nathlacroix commented:
Thanks for the pull request. @stephlazar pointed out rightfully that such a functionality exists in the analysis_toolbox (function called "latest_data()") although I wasn't aware of that when I wrote the function. We should probaly merge them and decide where we think is the best place for such function.

chellings commented:
Sorry, I did not know that. I could leave the pull request open, so that you could just commit the combined function to this branch and we can merge it to the qudev_master afterwards via this pull request. Or should I cancel the pull request?

nathlacroix commented:
I can do that yes. We should just maybe agree with @antsr and @stephlazar where we would like these types of functions in the future: in the analysis toolbox module which is a bit messy at the moment or if we want to integrate them as static functions of base classes. 
By the way, I think @stephlazar also started a larger clean-up of the analysis toolbox module, is that correct ?

chellings committed:
analysis_toolbox: improved method latest_data()
- option to return a list of n timestamps
- option to not return paths
- stop search faster if no more matches can be found due to newer_than
- changed return value for return_all=True to have same format as without this option
- some cleanup

chellings committed:
Revert "Add functionality to base analysis"

This reverts commit a7c835b2574ef1c2a5e1b4be134844c622211776.

chellings commented:
I have committed a proposal how to combine the functions, keeping it in analysis_toolbox for now. My proposed changes are backwards-compatible since they are based on optional arguments. Exception: output format in case of return_all=True, which I adapted as proposed by @stephlazar in her analysis cleanup.

stephlazar commented:
Yes, I created the pull request #32 some time ago which cleans up analysis_toolbox. I think it just deletes unused functions.

stephlazar commented on pycqed/analysis/analysis_toolbox.py:
```diff
@@ -118,47 +118,64 @@ def return_last_n_timestamps(n, contains=''):
+            raise_exc: Return False instead of raising an execption if no data is
+                found. (default: False)
+            folder: search directory (default: the stored datadir)
+            list_timestamps: If set to an integer n, a list of the n latest matches
```
For nicer code, I would suggest to rename this parameter to something like "num_ts", "n_ts", or just "n". In any case something that alludes to the fact that this parameter specified latest n timestamps. list_timestamps suggests that a list should be passed or that this parameter is a boolean flag. And then I would find it nicer to have its default value to 1 for the latest timestamp, instead of False.

chellings replied:
Calling it num_ts would also be fine for me. The reason for the default value is compatibility with old code, which will expect a single path/timestamp as return value and not a single-entry list. On the other hand, new code that uses num_ts will expect a list as return value and it is more robust to then return a single-entry list in case of num_ts=1 in order to have a unified return format. That is why I thought the distinction between n_ts not provided and n_ts=1 might be useful. What do you think? Of course, I could use None instead of False to encode the default behavior.

stephlazar replied:
The way I thought about it is that the default use case of this function remains to return the very last timestamp (ie num_ts = 1 by default, function returns a single path/timestamp, not a single-entry list). Only if the user specifies num_ts > 1 will this function return a list. Isn't that what you had in mind?

nathlacroix replied:
I would personally be ok with only the unified return format of lists. If you expect single answer, you can always take the first element of the list. In anycase, I think "n" or could make more sense as if I understood correctly, it's not only timestamps but also filepaths so the "list_timestamps" is a bit a misnomer, right?

stephlazar replied:
Yes I agree about it being a misnomer. Let's just cal lit "last_n." Then it's general it doesn't matter if the function returns timestamps or paths.

I we make this function always return lists, then this would break backwards compatibility and we would need to go change all the places in pycqed where this is used (including probably several notebooks that each have). So I wouldn't do this. Again, I'm not sure how you guys thought about the standard use case for this function. Because for me the scope of this function is primarily to return the very last data folder/timestamp, and rarely, when we need it for ex for your qaoa application, it can return a list of the several folders/timestamps

stephlazar commented:
Have you tested that this doesn't break our existing analyses? You could just call a Rabi or Ramsey analysis to see that it works. Since the latest_data function is used by the BaseAnalysis class to extract data.

chellings commented:
Test with Rabi/Ramsey analysis successful.

nathlacroix commented on pycqed/analysis/analysis_toolbox.py:
```diff
@@ -118,47 +118,64 @@ def return_last_n_timestamps(n, contains=''):
 
 
 def latest_data(contains='', older_than=None, newer_than=None, or_equal=False,
```
To try to create consistency with other similar functions I would maybe rename the function to "get_latest_data" (same naming convention as get_timestamps_in_range).

nathlacroix commented on pycqed/analysis/analysis_toolbox.py:
```diff
@@ -118,47 +118,64 @@ def return_last_n_timestamps(n, contains=''):
-        the requirements (Cristian)
-    '''
-    if (folder is None):
+                return_timestamp=False, return_path=True, raise_exc=True,
```
do we really want raise_exc to default to True? I would default it to false and return an empty list if nothing is found. If you explicitely want to raise an error because you're using it in a context where you require outputs then you could activate the flag.

stephlazar replied:
I think you would usually require outputs if you call this function, no? Is there a case where you wouldn't?

nathlacroix reviewed (changes requested):
Thanks for the pull request! I think minor changes could be useful (see comments).

chellings committed:
analysis_toolbox.latest_data: renamed list_timestamps to n_matches and changed default value to None
The value None indicates that the user has not provided a desired length of a list of matches, and the function will return a single item in this case. This is different from setting n_matches=1, in which case a one-element list will be returned.

chellings committed:
Merge branch 'qudev_master' into devel/get_latest_data

chellings committed:
analysis_toolbox.latest_data: fixed typos in docstring

chellings commented:
I have now changed the name of the argument to n_matches. The default behavior of returning a single item (not a list), which is used in existing code, is now encoded by the default value n_matches=None.

Justification: At the current stage, I would prefer having a change that is backwards compatible. If someone wants to do further changes and refactor the code at all points, where the function is used, this could of course be done in addition later on.

Changes that I have not implemented since they would require refactoring and/or additional testing of existing code:
- rename the function
- change default behavior to returning a single-entry list (n_matches=1)
- change default behavior to not raise an exception (raise_exc=False)

Please let me know if you can approve the pull request with this justification.

chellings committed:
analysis_toolbox.latest_data: small fixes for improved handling of some pathological cases

chellings commented:
Nathan's function always returned a list, even in case of a single timestamp. For example, if you write code that reads all n rows of a QAOA landscape (from n timestamps), you want to be able to loop over the rows (over the entries in a list), no matter which value n has (including n=1). You do not want to treat the case n=1 separately.

Since my task was to merge the functions, I implemented it in a way that both behaviors can be chosen by the caller. Setting n_matches=None will give the behavior of the old latest_data(). Setting n_matches to an integer (including n_matches=1) will give the behavior of Nathan's get_latest_data().


Thank you for pointing out that the caller might give inconsistent input. I have added an assertion ensuring that at least one of the two (return_timestamp or return_path) is True. Expected input is that either one of them si true or both are true.


Thank you to both of you for your reviews and for your thoughts about possible future changes.
